File: src/main/java/ua/softgroup/matrix/server/supervisor/producer/json/tracking/GeneralWorkDataJson.java
package ua.softgroup.matrix.server.supervisor.producer.json.tracking;

import com.fasterxml.jackson.annotation.JsonView;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

/**
 * @author Oleksandr Tyshkovets <sg.olexander@gmail.com>
 */
public class GeneralWorkDataJson {

    @JsonView(TrackingDataViewType.USER.class)
    private Long userId;

    @JsonView(TrackingDataViewType.PROJECT.class)
    private Long entityId;
    @JsonView(TrackingDataViewType.PROJECT.class)
    private String entityType = "project";

    @JsonView(TrackingDataViewType.DATE.class)
    private LocalDate date;

    private LocalTime start;
    private LocalTime end;
    private int dayWorkTimeSeconds;
    private List<TrackingPeriodJson> periods;

    public GeneralWorkDataJson() {
    }

    public GeneralWorkDataJson(LocalDate date, LocalTime start, LocalTime end, int dayWorkTimeSeconds,
                               List<TrackingPeriodJson> periods) {
        this.date = date;
        this.start = start;
        this.end = end;
        this.dayWorkTimeSeconds = dayWorkTimeSeconds;
        this.periods = periods;
    }

    public GeneralWorkDataJson(Long userId, LocalTime start, LocalTime end, int dayWorkTimeSeconds,
                               List<TrackingPeriodJson> periods) {
        this.userId = userId;
        this.start = start;
        this.end = end;
        this.dayWorkTimeSeconds = dayWorkTimeSeconds;
        this.periods = periods;
    }

    public GeneralWorkDataJson(Long entityId, String entityType, LocalTime start, LocalTime end,
                               int dayWorkTimeSeconds, List<TrackingPeriodJson> periods) {
        this.entityId = entityId;
//        this.entityType = entityType;
        this.start = start;
        this.end = end;
        this.dayWorkTimeSeconds = dayWorkTimeSeconds;
        this.periods = periods;
    }

    public Long getUserId() {
        return userId;
    }

    public void setUserId(Long userId) {
        this.userId = userId;
    }

    public Long getEntityId() {
        return entityId;
    }

    public void setEntityId(Long entityId) {
        this.entityId = entityId;
    }

    public String getEntityType() {
        return entityType;
    }

    public void setEntityType(String entityType) {
        this.entityType = entityType;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public LocalTime getStart() {
        return start;
    }

    public void setStart(LocalTime start) {
        this.start = start;
    }

    public LocalTime getEnd() {
        return end;
    }

    public void setEnd(LocalTime end) {
        this.end = end;
    }

    public int getDayWorkTimeSeconds() {
        return dayWorkTimeSeconds;
    }

    public void setDayWorkTimeSeconds(int dayWorkTimeSeconds) {
        this.dayWorkTimeSeconds = dayWorkTimeSeconds;
    }

    public List<TrackingPeriodJson> getPeriods() {
        return periods;
    }

    public void setPeriods(List<TrackingPeriodJson> periods) {
        this.periods = periods;
    }
}

File: src/main/scala/ua/softgroup/matrix/server/supervisor/producer/json/tracking/GeneralWorkDataJson.scala
package ua.softgroup.matrix.server.supervisor.producer.json.tracking

import java.time.{LocalDate, LocalTime}

import com.fasterxml.jackson.annotation.JsonView

import scala.beans.BeanProperty

/**
  * @author Oleksandr Tyshkovets <olexandr.tyshkovets@gmail.com>
  */
class GeneralWorkDataJson {

  @JsonView(Array(classOf[TrackingDataViewType.USER]))
  @BeanProperty
  var userId: Long = _

  @JsonView(Array(classOf[TrackingDataViewType.PROJECT]))
  @BeanProperty
  var entityId: Long = _
  @JsonView(Array(classOf[TrackingDataViewType.PROJECT]))
  @BeanProperty
  var entityType: String = "project"

  @JsonView(Array(classOf[TrackingDataViewType.DATE]))
  @BeanProperty
  var date: LocalDate = _

  @BeanProperty
  var start: LocalTime = _
  @BeanProperty
  var end: LocalTime = _
  @BeanProperty
  var dayWorkTimeSeconds: Int = _
  @BeanProperty
  var periods: List[TrackingPeriodJson] = _

  def this(date: LocalDate, start: LocalTime, end: LocalTime, dayWorkTimeSeconds: Int,
           periods: List[TrackingPeriodJson]) {
    this()
    this.date = date
    this.start = start
    this.end = end
    this.dayWorkTimeSeconds = dayWorkTimeSeconds
    this.periods = periods
  }

  def this(userId: Long, start: LocalTime, end: LocalTime, dayWorkTimeSeconds: Int,
           periods: List[TrackingPeriodJson]) {
    this()
    this.userId = userId
    this.start = start
    this.end = end
    this.dayWorkTimeSeconds = dayWorkTimeSeconds
    this.periods = periods
  }

  def this(entityId: Long, entityType: String, start: LocalTime, end: LocalTime,
           dayWorkTimeSeconds: Int, periods: List[TrackingPeriodJson]) {
    this()
    this.entityId = entityId
//    this.entityType = entityType
    this.start = start
    this.end = end
    this.dayWorkTimeSeconds = dayWorkTimeSeconds
    this.periods = periods
  }

  override def toString: String = {
    "GeneralWorkDataJson(userId=%d, entityId=%d, entityType=%s, date=%s, start=%s, end=%s, dayWorkTimeSeconds=%d, periods=%s)"
      .format(userId, entityId, entityType, date, start, end, dayWorkTimeSeconds, periods)
  }

}
